akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -693,3 +699,470 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+    private var pending = Vector.empty[T]
+    private var consumers: Vector[Consumer] = Vector.empty
+    private val needWakeup: LongMap[Consumer] = LongMap.empty
+    private var consumerIds: Array[Long] = Array.emptyLongArray
+
+    private var callbackCount = 0L
+
+    override def preStart(): Unit = {
+      setKeepGoing(true)
+      callbackPromise.success(getAsyncCallback[HubEvent](onEvent))
+      if (startAfterNbrOfConsumers == 0)
+        pull(in)
+    }
+
+    override def onPush(): Unit = {
+      publish(grab(in))
+      if (!isFull) pull(in)
+    }
+
+    private def isFull: Boolean = {
+      (queue.size + pending.size) >= bufferSize
+    }
+
+    private def publish(elem: T): Unit = {
+      if (!initialized || consumers.isEmpty) {
+        // will be published when first consumers are registered
+        pending :+= elem
+      } else {
+        val id = materializedPartitioner(consumerIds, elem)
+        queue.offer(id, elem)
+        wakeup(id)
+      }
+    }
+
+    private def wakeup(id: Long): Unit = {
+      needWakeup.get(id) match {
+        case None ⇒
+        case Some(consumer) ⇒
+          needWakeup -= id
+          consumer.callback.invoke(Wakeup)
+      }
+    }
+
+    override def onUpstreamFinish(): Unit = {
+      if (consumers.isEmpty)
+        completeStage()
+      else {
+        consumers.foreach(c ⇒ complete(c.id))
+      }
+    }
+
+    private def complete(id: Long): Unit = {
+      queue.offer(id, Completed)
+      wakeup(id)
+    }
+
+    private def tryPull(): Unit = {
+      if (initialized && !isClosed(in) && !hasBeenPulled(in) && !isFull)
+        pull(in)
+    }
+
+    private def onEvent(ev: HubEvent): Unit = {
+      callbackCount += 1
+      ev match {
+        case NeedWakeup(consumer) ⇒
+          // Also check if the consumer is now unblocked since we published an element since it went asleep.
+          if (queue.nonEmpty(consumer.id))
+            consumer.callback.invoke(Wakeup)
+          else {
+            needWakeup.update(consumer.id, consumer)
+            tryPull()
+          }
+
+        case TryPull ⇒
+          tryPull()
+
+        case RegistrationPending ⇒
+          state.getAndSet(noRegistrationsState).asInstanceOf[Open].registrations foreach { consumer ⇒
+            consumers :+= consumer
+            consumers = consumers.sortBy(_.id)
+            consumerIds :+= consumer.id
+            consumerIds.sorted
+            queue.init(consumer.id)
+            if (consumers.size >= startAfterNbrOfConsumers) {
+              initialized = true
+            }
+
+            consumer.callback.invoke(Initialize)
+
+            if (initialized && pending.nonEmpty) {
+              pending.foreach(publish)
+              pending = Vector.empty[T]
+            }
+
+            tryPull()
+          }
+
+        case UnRegister(id) ⇒
+          consumers = consumers.filterNot(_.id == id)
+          consumerIds = consumerIds.filterNot(_ == id)
+          queue.remove(id)
```

Comment:
That is no different than if the elements were in flight further downstream.
Rearrange would be weird since element order would be violated.